### PR TITLE
Move init_ddp_connection to DDP Plugin

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This moves `init_ddp_connection` to the DDP Plugin, making it easily overridable in the same place as the `configure_ddp` for the model. This means custom implementations can be shared easily cross accelerator backends

Some questions:
- Should this live in the specific DDP accelerators? i made this change to avoid copy/paste across accelerators by stick everything in the base class and to get feedback. DDP doesn't apply in the base class so it sticks out right now
- On the other hand, that copy/paste already exists for `configure_ddp`

cc @williamFalcon  @awaelchli 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
Do we need the `init_ddp_connection` in the accelerator at all if it just forwards to the plugin? Or is there anything different across different types of DDP?

**Comment by ananthsub** [comment]
@justusschock `init_ddp_connection` used to be a hook on the lightning module, then it was moved to the accelerator. there are use cases which want to customize this (e.g. create custom process groups), but currently they'd need to write a whole accelerator backend just to do so

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 1]
```diff
@@ -62,3 +63,24 @@ def configure_ddp(self, model, device_ids):
...
         )
         return model
+
+    def init_ddp_connection(
+        self,
+        trainer,
+        cluster_environment,
+        global_rank: int,
+        world_size: int,
+        is_slurm_managing_tasks: bool = True,
```
Comment (ananthsub) [inline pytorch_lightning/plugins/ddp_plugin.py:73]: i don't know if this is the right interface. is this leaking too much? should the plugin have a reference to the trainer?

Comment (justusschock) [reply pytorch_lightning/plugins/ddp_plugin.py:73]: IMO we already have too many things with trainer references.  Besides that I feel that this is fine.

Comment (williamFalcon) [reply pytorch_lightning/plugins/ddp_plugin.py:73]: as mentioned earlier, we’re cleaning up trainer references for the next stage of refactoring. 

but i think that’s for a different PR.

**Review by awaelchli — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

**Pull request merged**